### PR TITLE
Support ODBC keyword CURRENTSCHEMA

### LINKED ISSUE #578: Failed to get current schema when using IBM Cloud IAM authentication
## Problem

environment: Python 3.7, ibm_db 3.0.2

When following the official guide [Identity and access management (IAM) on IBM Cloud](https://www.ibm.com/support/knowledgecenter/en/SS6NHC/com.ibm.swg.im.dashdb.security.doc/doc/iam.html)
Given Db2 warehouse in cloud, 
And example file `test_sqlalchemy.py` like this:
```Python
from sqlalchemy import text
from sqlalchemy import create_engine

SQLALCHEMY_URL= f'db2+ibm_db://@{db_host}:50001/BLUDB;CURRENTSCHEMA=KZHU_TEST;PROTOCOL=TCPIP;SECURITY=SSL;Authentication=GSSplugin;AccessToken={JWT_TOKEN}'

sql = text('SELECT * FROM TEST_ACTOR')
engine = create_engine(os.environ['SQLALCHEMY_URL'])
result = engine.execute(sql)

names = [row[0] for row in result]

print(names)
```

It will throw this error when running above python code:

```
pipenv run python test_sqlalchemy.py 
Loading .env environment variables...
user:  
Traceback (most recent call last):
  File "test_sqlalchemy.py", line 12, in <module>
    result = engine.execute(sql)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2234, in execute
    connection = self._contextual_connect(close_with_result=True)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2302, in _contextual_connect
    self._wrap_pool_connect(self.pool.connect, None),
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2336, in _wrap_pool_connect
    return fn()
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 364, in connect
    return _ConnectionFairy._checkout(self)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 778, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 495, in checkout
    rec = pool._do_get()
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/impl.py", line 140, in _do_get
    self._dec_overflow()
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    with_traceback=exc_tb,
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 182, in raise_
    raise exception
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/impl.py", line 137, in _do_get
    return self._create_connection()
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 309, in _create_connection
    return _ConnectionRecord(self)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 440, in __init__
    self.__connect(first_connect_check=True)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 666, in __connect
    ).exec_once_unless_exception(self.connection, self)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/event/attr.py", line 314, in exec_once_unless_exception
    self._exec_once_impl(True, *args, **kw)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/event/attr.py", line 285, in _exec_once_impl
    self(*args, **kw)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/event/attr.py", line 322, in __call__
    fn(*args, **kw)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 1397, in go
    return once_fn(*arg, **kw)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py", line 199, in first_connect
    dialect.initialize(c)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/ibm_db_sa/base.py", line 698, in initialize
    super(DB2Dialect, self).initialize(connection)
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 319, in initialize
    connection
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/ibm_db_sa/ibm_db.py", line 202, in _get_default_schema_name
    return self.normalize_name(connection.connection.get_current_schema())
  File "/root/.local/share/virtualenvs/butane-c47Aapyx/lib/python3.7/site-packages/ibm_db_dbi.py", line 812, in get_current_schema
    return self.current_schema
AttributeError: 'Connection' object has no attribute 'current_schema'
```

## Analysis

First of all, a similar DSN works when using JDBC interface, so the IAM authentication works as designed.

Notice that there's NO user id nor password since the credential will be passed via the access token which includes such info with signature. So the below line of code will cause the current_schema to be an empty string:

https://github.com/ibmdb/python-ibmdb/blob/778f63228517040ca5c3d4c3d77589116c635c9e/IBM_DB/ibm_db/ibm_db_dbi.py#L600

When retrieving current schema, it just ignored the current schema since it's an empty string:

https://github.com/ibmdb/python-ibmdb/blob/778f63228517040ca5c3d4c3d77589116c635c9e/IBM_DB/ibm_db/ibm_db_dbi.py#L801-L810

Then the above AttributeError is expected since there indeed is no attribute named `current_schema`.

## Suggested fix

It could introduce a new feature to allow overriding the current_schema via ODBC option `CURRENTSCHEMA`, so that if there's no userid, it still work if user provides this option.

### PR DESCRIPTION
**Pull request by shawnzhu**
Fixes #578

### CODE REVIEW

File: IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        self.run_default_current_schema()
+        self.run_setup_schema()
+
+    def _assert_current_schema(self, conn, schema_name):
+        query = "SELECT CURRENT_SCHEMA FROM SYSIBM.SYSDUMMY1"
+        try:
+            stmt = ibm_db.exec_immediate(conn, query)
+            res = ibm_db.fetch_tuple(stmt)
+            self.assertEqual(res[0].strip(), schema_name.upper())
+        except:
```
Comment (meneal) [inline IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py:26]: I would suggest something slightly different here. I assume the point of the except is to capture some query error or something like it with the db. If that is the case then I would suggest:

```
except Exception as e:
    print("Query failed for: ", e)
    raise
```

It results in the same thing, it just doesn't obscure what the error was in the case that some debugging needs to happen.

File: IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            self.assertIsNone(sys.exc_info()[0], "Query failed")
+
+    def run_default_current_schema(self):
+        conn = ibm_db.connect(config.database, config.user, config.password)
+
+        if conn:
+            self._assert_current_schema(conn, config.user)
+
+        else:
+            print ("Connection failed.")
```
[the remark this answers is not in the document]
Comment (meneal) [reply IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py:36]: Is this else needed? The reason I ask is that it appears that `ibm_db.connect` will throw an error if there is a connection failure. Is that not the case?

File: IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    def run_setup_schema(self):
+        conn = ibm_db_dbi.connect(
+            'CURRENTSCHEMA=SYSIBM;PORT=%d;PROTOCOL=TCPIP;' % config.port,
+            config.user, config.password, config.hostname, config.database
+        )
+
+        if conn:
+            self.assertEqual(conn.get_current_schema(), 'SYSIBM')
+            self._assert_current_schema(conn.conn_handler, 'SYSIBM')
```
Comment (meneal) [inline IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py:47]: Why use both `conn.get_current_schema()` and `_assert_current_schema`? Wouldn't one test the same thing that the other does?

**Review by meneal — COMMENTED** [review]
A few questions and a suggestion. Thanks for your work on this!

File: IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py (hunk 1 above)
Comment (shawnzhu) [reply IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py:26]: Checkout https://stackoverflow.com/questions/4990718/about-catching-any-exception

If you insistent your suggestion, create a PR to the origin repo :)

File: IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shawnzhu) [reply IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py:36]: This is the linting rule: https://github.com/ibmdb/python-ibmdb/blob/778f63228517040ca5c3d4c3d77589116c635c9e/.travis.yml#L16

and you can always fix such problem via PR to the origin if you like.

File: IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py (hunk 3 above)
Comment (shawnzhu) [reply IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py:47]: No, they are two assertions:

1. first one is testing the current schema is set up as meta info for the `Connection` object (OUT of the actual ODBC driver connection handle)
2. the second one is testing it does set up the current schema WITHIN the execution context of the ODBC driver connection handle.

File: IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shawnzhu) [reply IBM_DB/ibm_db/ibm_db_tests/test_currentschema.py:36]: > Is this else needed?

This is a good question for all the test code in this repo which is out of the scope of this PR.

**Pull request merged**